<h2 align='center' style='color:blue'>Dropout Regularization In Deep Neural Network</h2>

This is a dataset that describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It is a **binary classification problem** that requires a model to differentiate rocks from metal cylinders.

Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks)
Download it from here: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('D:\\6th sem\\deep learning\\sonar_dataset.csv', header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
185  0.0340  0.0625  0.0381  0.0257  0.0441  0.1027  0.1287  0.1850  0.2647   
190  0.0156  0.0210  0.0282  0.0596  0.0462  0.0779  0.1365  0.0780  0.1038   
149  0.0207  0.0535  0.0334  0.0818  0.0740  0.0324  0.0918  0.1070  0.1553   
115  0.0414  0.0436  0.0447  0.0844  0.0419  0.1215  0.2002  0.1516  0.0818   
139  0.0164  0.0627  0.0738  0.0608  0.0233  0.1048  0.1338  0.0644  0.1522   

         9   ...      51      52      53      54      55      56      57  \
185  0.4117  ...  0.0141  0.0019  0.0067  0.0099  0.0042  0.0057  0.0051   
190  0.1567  ...  0.0150  0.0060  0.0082  0.0091  0.0038  0.0056  0.0056   
149  0.1234  ...  0.0033  0.0050  0.0190  0.0103  0.0121  0.0042  0.0090   
115  0.1975  ...  0.0045  0.0136  0.0113  0.0053  0.0165  0.0141  0.0077   
139  0.0780  ...  0.0258  0.0143  0.0226  0.0187  0.0185  0.0110  0.0094   

         58      59  60  
185  0.0033  0.0058   M  
190  0.0048  0.0024   M  
149  0.0070  0.0099   M  
115  0.0246  0.0198   M  
139  0.0078  0.0112   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60],
      dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

60
M    111
R     97
Name: count, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
24    True
124  False
199  False
190  False
197  False

In [10]:
y.value_counts()

R    
False    111
True      97
Name: count, dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

# Using Deep Learning Model

## Model without Dropout Layer

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 3s 5ms/step - loss: 0.6965 - accuracy: 0.4615
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6638 - accuracy: 0.6346
Epoch 3/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6375 - accuracy: 0.7244
Epoch 4/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6431 - accuracy: 0.6282
Epoch 5/100
20/20 [==============================] - 0s 5ms/step - loss: 0.5918 - accuracy: 0.7308
Epoch 6/100
20/20 [==============================] - 0s 4ms/step - loss: 0.5614 - accuracy: 0.7756
Epoch 7/100
20/20 [==============================] - 0s 4ms/step - loss: 0.5259 - accuracy: 0.7564
Epoch 8/100
20/20 [==============================] - 0s 4ms/step - loss: 0.4915 - accuracy: 0.8013
Epoch 9/100
20/20 [==============================] - 0s 4ms/step - loss: 0.4663 - accuracy: 0.8077
Epoch 10/100
20/20 [==============================] - 0s 3ms/step - loss: 0.4376 - accuracy: 0.8269
Epoch 11/

In [16]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 10ms/step - loss: 0.8749 - accuracy: 0.7885


[0.8749132752418518, 0.7884615659713745]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 8ms/step
[4.5559791e-06 8.8200146e-01 9.9576050e-01 2.0691846e-04 1.0000000e+00
 9.9999398e-01 1.4519073e-02 1.0000000e+00 6.1882740e-05 1.0000000e+00]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  False
155  False
165  False
200  False
58    True
34    True
151  False
18    True
202  False
62    True

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.75      0.89      0.81        27
        True       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52



### Model with Dropout Layer

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 3s 5ms/step - loss: 0.7585 - accuracy: 0.4808
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7114 - accuracy: 0.5192
Epoch 3/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7049 - accuracy: 0.5192
Epoch 4/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6743 - accuracy: 0.5641
Epoch 5/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6926 - accuracy: 0.5449
Epoch 6/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6871 - accuracy: 0.5449
Epoch 7/100
20/20 [==============================] - 0s 5ms/step - loss: 0.7136 - accuracy: 0.5385
Epoch 8/100
20/20 [==============================] - 0s 4ms/step - loss: 0.7047 - accuracy: 0.5256
Epoch 9/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6844 - accuracy: 0.5256
Epoch 10/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6858 - accuracy: 0.5000
Epoch 11/

In [21]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 13ms/step - loss: 0.4770 - accuracy: 0.7692


[0.47702810168266296, 0.7692307829856873]

Training Accuracy is still good but Test Accuracy Improved

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 10ms/step
[1.4203839e-04 8.4864676e-01 9.1377735e-01 9.1611054e-03 9.9808258e-01
 9.0927207e-01 4.2302805e-01 9.9789453e-01 1.6948447e-02 9.9885708e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.73      0.89      0.80        27
        True       0.84      0.64      0.73        25

    accuracy                           0.77        52
   macro avg       0.78      0.76      0.76        52
weighted avg       0.78      0.77      0.77        52



**You can see that by using dropout layer test accuracy increased from 0.77 to 0.81**